### DISCUSSION TITLE
Need some help trying to bzlmod rules_ros2

### DISCUSSION THREAD

**Discussion by lalten** [mvukov/rules_ros2#235] [General]
I have some work in progress at https://github.com/lalten/rules_ros2/pull/1

I'm stuck on two issues:

* I don't understand why but Python 3.10 won't work with rules_python. The error message is `Error in fail: Unable to find interpreter for pip hub 'rules_ros2_pip_deps' for python_version=3.10: Make sure a corresponding python.toolchain(python_version="3.10") call exists`. Changing to 3.11 works totally fine.
* this line seems to be causing some issue: https://github.com/mvukov/rules_ros2/blob/8fedbc779b32e762f5f1de0962e1e0ac35cb8c56/ros2/ament.bzl#L207
```
ERROR: file 'ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg' is generated by these conflicting actions:
Label: //ros2/test/rosbag:bag_launch_ament_setup
RuleClass: ros2_ament_setup rule
JavaActionClass: class com.google.devtools.build.lib.analysis.actions.SymlinkAction
Configuration: 652a065232f76ebe9d5f97de3542e54d496b05434136fa383ba30bebfb4953dd
Mnemonic: Symlink
Action key: 7a39b61d[57](https://github.com/lalten/rules_ros2/actions/runs/7423337220/job/20200572229?pr=1#step:5:58)44a8903119e17b3ddc7e56e330f240d2bead98c3ba081a81b5c9ac
Progress message: Creating symlink bazel-out/k8-fastbuild/bin/ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg
Action describeKey: (null)
PrimaryInput: File:[/github/home/.cache/bazel/_bazel_runner/d53d6[63](https://github.com/lalten/rules_ros2/actions/runs/7423337220/job/20200572229?pr=1#step:5:64)888cb7d4fa0ce43a097f11d2b/external/_main~rules_ros2_repositories~ros2_rcl_interfaces[source]]builtin_interfaces/msg/Duration.msg, File:[/github/home/.cache/bazel/_bazel_runner/d53d663888cb7d4fa0ce43a097f11d2b/external/ros2_rcl_interfaces[source]]builtin_interfaces/msg/Duration.msg
PrimaryOutput: File:[[<execution_root>]bazel-out/k8-fastbuild/bin]ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg
ERROR: /__w/rules_ros2/rules_ros2/ros2/test/rosbag/BUILD.bazel:27:9: for ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg, previous action: action 'Creating symlink bazel-out/k8-fastbuild/bin/ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg', attempted action: action 'Creating symlink bazel-out/k8-fastbuild/bin/ros2/test/rosbag/bag_launch_ament_setup/share/builtin_interfaces/msg/Duration.msg'
```
 * there are probably more problems to be discovered after these ;)

Do you have any smart ideas @mvukov @ahans ? :)

**Comment by ziad-fernride** [comment]
for issue number 1, have you tried setting the is_default to true when using toolchain 3.10 ? it could be that the 3.11 from rules_python is picked since it is the first default found

**Reply by lalten** [reply]
Is it correct to set is_default in a ruleset? I thought it was for root modules only

**Reply by ziad-fernride** [reply]
I think rules_python does a breadth_first search and takes the first is_default it finds ... since the root is processed first in the search, setting is_default in the root should always make the root the decision maker

**Reply by ziad-fernride** [reply]
the behaviour should be clearer after this change is released: 
https://github.com/bazelbuild/rules_python/commit/619fa0c8a45fb33cb889d3105c6fbe049cfb3e50

**Comment by lalten** [comment]
I was able to solve the duplicate symlink issue. I had some duplication going on with WORKSPACE.bzlmod and the module extension bzl. Removed the module extension and left only the hybrid-mode WORKSPACE.bzlmod (See https://bazel.build/external/migration#hybrid-mode) for now, and the tets pass :)

**Comment by mvukov** [comment]
Great effort. Feel free to target the PR against this repo.

**Comment by mikebauer** [comment]
Let me start with a big disclaimer that I'm not a bazel expert and I've probably made some questionable decisions in how I'm going about this. I've been transitioning my company's open source repo (which depends on rules_ros2) to use bzlmod and in order to get it building I've quickly put together a registry containing the subset of rules_ros2 that my company's open source repo depends on:

https://github.com/resim-ai/open-core/pull/132

I admit this may be a bit of a strange way to solve this (i.e. just landing a custom registry into the repo itself), but I figured I would mention it on the off chance that it could be at all useful for this project. In particular, if you'd like me to push any of these changes upstream or to the central registry, I'd be happy to. On the other hand, I won't be offended if this looks like the wrong way to do things from y'all's perspective. Further disclaimers:
 * I haven't included all patches or all repos from rules_ros2 in this registry. It's limited to the ones I actually depend on at the moment.
 * I haven't self-reviewed the code yet so there may be metadata errors or missing transitive dependencies that my code doesn't actually rely on.
 * Some of the patches from rules_ros2 have been modified in accordance with my own needs (e.g. rmw_implementation_library_path.patch) and may be inapropriate for others.
